### DISCUSSION TITLE
Network card based on Raspberry Pi

### DISCUSSION THREAD

**Discussion by toncho11** [jbruchon/elks#1474] [Ideas]
@Mellvik I think this is for you!

I think the best option for a network card for ELKS will be a newly produced one. This can be done with a Raspberry PI.
Check: 

* https://www.youtube.com/watch?v=M7Lov_MH6n
* https://hackaday.com/2022/11/13/emulate-any-isa-card-with-a-raspberry-pi-and-an-fpga/

This will reduce the need to implement many network drivers for ELKS.

**Comment by toncho11** [comment]
Usually these projects work like that:

* a PCB is designed where the Pi is plugged (as well as the FPGA)
* a firmware is written for the Pi - either bare metal or for Linux on the Pi
* a driver is written for DOS/ELKS

Examples are sound cards and [floppy controllers](https://www.smbaker.com/raspberry-pi-virtual-floppy-for-isa-pc-xtat-computers).

**Comment by Mellvik** [comment]
Wow, Thank you @toncho11 - very interesting indeed. I'd love to 'play' with this one. Infinite possibilities...If you find a link to where such an ISA board can be purchased, let me know.-M7. des. 2022 kl. 22:45 skrev toncho11 ***@***.***>:﻿
Usually these projects work like that:

a PCB is designed where the Pi is plugged (as well as the FPGA)
a firmware is written for the Pi - either bare metal or for Linux on the Pi
a driver is written for DOS/ELKS

**Comment by toncho11** [comment]
Actually everything is already done. 

Check the https://www.smbaker.com/raspberry-pi-virtual-floppy-for-isa-pc-xtat-computers , watch the video. 

It uses a shared memory on a ISA card to communicate with the Raspberry Pi Zero. It emulates a floppy drive through a BIOS extension. I think one can modify it to emulate a network card. You just need to write a driver that places all network packets on the shared memory where the Raspberry Pi will simply forward them to your local network. The Pi Zero does not have a network port by default, but this is not a problem. The guy shows that you can use a standard Raspberry Pi.  

I think it is quite doable. The card is already done!

**Comment by toncho11** [comment]
So it seems it is open source on github except that ordering the PCB is a little bit on the expensive side. 

https://oshpark.com/shared_projects/x9tFlsBX

If you add 20$ (just guessing) for ICs and then the price of the RaspberryPi then it can go around 100-110$. 
I have attached the .brd file. With this file it might be cheaper to print the PCB with another company. 

[894f40854469a268b02cf90fefc4eabc.zip](https://github.com/jbruchon/elks/files/10184372/894f40854469a268b02cf90fefc4eabc.zip)

**Comment by Mellvik** [comment]
Thanks again @toncho11.

It's hard not to get sucked right into this and disconnect from the real world while dreaming up all kinds of interesting projects. I really like the idea of having a RaspberryPi available on the ISA bus. It's unfortunate that the shared memory is only 1k, expanding it to 2k would have made a world of difference.

stil, as is this design has a lot of potential for fun ELKS projects - using the Raspi Zero or ZeroW. The Zero (without the builtin WiFi) can run ethernet-on-usb which works out of the box with MacOS. And of course we can use a USBdongle with WiF or wired Ethernet. I actually have an unused ZeroW sitting in  front of me right now. Used to be a FM transmitter (internet Radio -> FM) for the neighbourhood since FM was replaced by DAB around my neck of the woods. Giga-bummer, yet here we are … 

YEs, the card is really expensive and I hate soldering surface mount components, but it's doable. Maybe 2023 … Tempting indeed.

—M

**Comment by toncho11** [comment]
It will resolve the current problem where you search for a specific old card or close one and then you hope that the existing driver will work or worse you need to write one using scarce documentation. But then as you said you need to solder it ...
